Reload Mods button changes and available mod updates indicator

jackbondpreston opened the pull request:
### What is the new feature?
Aims to address #1386

- "Reload Mods" button renamed to "Force Reload All"
- "Force Reload All" button only shows when in developer mode
- New text on the "Mod Browser" entry on main menu to indicate number of available mod updates

### Why should this be part of tModLoader?
-  Button changes were suggested in Discord - I believe this reload button shouldn't really be useful to non-dev users anymore?
- Update reminder is useful to prevent users forgetting to update mods. This will still be useful post-steam-workshop for GOG users.

### Are there alternative designs?
An icon could be used for indicating available mod updates - either some !-like indicator or perhaps a number in a circle. But I wasn't sure how to do this, and the text is clearer in meaning.

### Screenshots
- Mod Update Available
![image](https://user-images.githubusercontent.com/1207989/117599806-827b2b00-b142-11eb-8134-f277e18c1a74.png)
![image](https://user-images.githubusercontent.com/1207989/117600696-5791d680-b144-11eb-9689-b3cd63babf24.png)

- Mods Menu (Developer Mode = True)
![image](https://user-images.githubusercontent.com/1207989/117600835-a8a1ca80-b144-11eb-8d3b-3876530471fd.png)

- Mods Menu (Developer Mode = False)
![image](https://user-images.githubusercontent.com/1207989/117600529-01bd2e80-b144-11eb-8955-6986342eb560.png)

Commit:
Rename reload mods button and only show in developer mode

Rename "Reload Mods" -> "Force Reload All"
Hide the force reload button when not in developer mode, and adjust layout accordingly

Commit:
Use new localisation key instead of changing previous one

Commit:
Add indicator for available mod updates in main menu

Commit:
Merge branch 'tModLoader:1.4' into 1.4

Commit:
Rename "Force Reload All" -> "Force Reload"

Commit:
Didn't mean to change the previous ModsReloadMods string

Solxanich reviewed (commented):
Just a quick formatting review.

Commit:
Improvements/fixes based on review

jackbondpreston commented on patches/tModLoader/Terraria/ModLoader/ModLoader.cs:
```diff
@@ -195,6 +202,86 @@ private static void Load(CancellationToken token = default)
+			}
+			catch (Exception e) {
+				// Fail silently, it doesn't really matter if the update indicator is missing
+				Logging.tML.Debug("Error checking for mod updates.");
```
Is localisation necessary for debug logs? I couldn't find any existing examples of these having been localisated.

Chicken-Bones replied:
We're a bit inconsistent on these ones. I'd say it's not necessary outside of ModCompile

direwolf420 commented:
I'm not a friend of the main menu having a "YOU DON'T HAVE X AMOUNT OF MODS UPDATED YOU FOOL" in the middle of the screen, but that's just me.

Steviegt6 commented:
> I'm not a big fan of the main menu having a "YOU DON'T HAVE X AMOUNT OF MODS UPDATED YOU FOOL" in the middle of the screen, but that's just me.

+1
The original proposal's design was incredibly more elegant in concept. I don't feel like we need "(x Update(s) Available) in the middle of our screen acting as an eye sore just for the sake of it. Drawing a little symbol wouldn't be challenging (would be even easier if my Menu Buttons PR was merged, SMH), or you could just change the color or something and give the button some mouse text on hover. Not sure of what would look the best, but the current design stands out in a bad way.

jackbondpreston commented:
I'm happy to give the original proposal a shot if anyone can give guidance on where/how to go about drawing an icon overlaying the main menu.

Re: locking reload mods, it was discussed on discord and suggested to do this - I'm not totally in the know about this but I think it seemed like mods are automatically reloaded like you mentioned? Someone else can probably shine light on this, I'm happy to revert it if not.

Solxanich commented:
Highlights from Discord for reload button:
1) mods are auto-reloaded when leaving the menu if there are changes already. 
2) The button is thus a Force Reload button
3) Does anyone know why a player would need to Force Reload? No, we guess only modders have use of this, so hide it behind Developer mode

jackbondpreston commented:
I'll take a look at trying the redesign/check out your PR later today as I agree it would be better.

In terms of removing the button, I don't mind either way but I'll let people more knowledgeable than me make that call - feel free to give any opinions on that

Steviegt6 commented:
> I'll take a look at trying the redesign/check out your PR later today as I agree it would be better.
> 
> In terms of removing the button, I don't mind either way but I'll let people more knowledgeable than me make that call - feel free to give any opinions on that

I think the discussion should be brought up in Discord again while this time keeping in mind that normally you can enable and disable automatic reloading, and there are actual uses in terms of saving time, hassle, and resouces.

Commit:
Code style improvements

Mirsario commented on patches/tModLoader/Terraria/ModLoader/ModLoader.cs:
```diff
@@ -195,6 +202,86 @@ private static void Load(CancellationToken token = default)
+
+			try {
+				Task.Factory.StartNew(() => {
+					ServicePointManager.Expect100Continue = false;
+					const string url = "http://javid.ddns.net/tModLoader/listmods.php";
+					var values = new NameValueCollection {
+						{"modloaderversion", BuildInfo.versionedName},
+						{"platform", ModLoader.CompressedPlatformRepresentation},
+						{"netversion", FrameworkVersion.Version.ToString()},
+						{"EarlyAutoUpdate", UI.ModBrowser.UIModBrowser.EarlyAutoUpdate.ToString()}
+					};
+					using (var client = new WebClient()) {
+						ServicePointManager.ServerCertificateValidationCallback = (sender, certificate, chain, policyErrors) => { return true; };
+						client.UploadValuesCompleted += UpdateCheckComplete;
+						client.UploadValuesAsync(new Uri(url), "POST", values);
+					}
```
```suggestion
					ServicePointManager.Expect100Continue = false;
					
					const string Url = "http://javid.ddns.net/tModLoader/listmods.php";
					
					var values = new NameValueCollection {
						{ "modloaderversion", BuildInfo.versionedName },
						{ "platform", ModLoader.CompressedPlatformRepresentation },
						{ "netversion", FrameworkVersion.Version.ToString() },
						{ "EarlyAutoUpdate", UI.ModBrowser.UIModBrowser.EarlyAutoUpdate.ToString() }
					};
					
					using var client = new WebClient();
					
					ServicePointManager.ServerCertificateValidationCallback = (sender, certificate, chain, policyErrors) => true;
					client.UploadValuesCompleted += UpdateCheckComplete;
					
					client.UploadValuesAsync(new Uri(Url), "POST", values);
```

Mirsario reviewed (changes requested):
I think it would be better to display the available mod updates to the bottom left corner of the screen instead of having them on the mod browser button (which doesn't even exist anymore? Merge upstream.)

Show `X Mods Enabled (Y Updates Available!)` below TML's version info and patreon link.

Mirsario committed:
Update patches/tModLoader/Terraria/ModLoader/ModLoader.cs

Mirsario committed:
Update patches/tModLoader/Terraria/ModLoader/ModLoader.cs

Commit:
Merge branch 'tModLoader:1.4' into 1.4

Commit:
Move update indicator to bottom left of main menu

Commit:
Remove newline as I am no longer modifying this file

jackbondpreston commented:
> 2\. Not everyone wants to immediately reload their mods, especially when they have larger mods enabled. I find myself often wanting to not immediately reload if I have to update mods or have others I wish to install (entirely separate from when I develop mods).
>     I don't see any reason to get rid of a perfectly-functioning feature.

How would you feel about a setting which enables auto reloading mods which require a reload upon exiting the Mods screen? When disabled the Force Reload button could be shown instead.

Commit:
Prevent enabled mods count from going below 0 (this could happen whilst reloading)

Commit:
Add functionality to auto reload required mods

jackbondpreston commented:
Ok, I have pushed the functionality I described above:
New config option `AutomaticallyReloadRequiredModsLeavingModsScreen` which enables/disables automatically reloading mods that require a reload when the user exits the mod screen. If this option is disabled, the Force Reload button will be displayed. If it is enabled, the button is hidden.

Commit:
Merge remote-tracking branch 'upstream/1.4' into 1.4

Commit:
Move update checking logic outside of ModLoader.cs

Commit:
Merge remote-tracking branch 'upstream/1.4' into 1.4

jackbondpreston commented:
Hopefully I addressed your suggestions @Mirsario.

I guess it would be best if this change only checks for mod updates and displays the mod updates available text when a GOG instance is running (the x mods loaded could stay for Steam)?

Solxanich committed:
Merge branch '1.4' into pr/1516

Solxanich committed:
Cleanup files after merge conflicts

Solxanich committed:
Get the PR working again
